### PR TITLE
MotoROS v1.8.1

### PR DESCRIPTION
**Pull request by gavanderhoorn**
See commit comments.

### CODE REVIEW

**Comment by gavanderhoorn** [comment]
@EricMarcil: fyi.

**Comment by gavanderhoorn** [comment]
Note that we'll need to update the links on the wiki again after this gets merged (because of the new binaries). See #220.

File: motoman_driver/MotoPlus/CtrlGroup.c
Comment (gavanderhoorn) [inline motoman_driver/MotoPlus/CtrlGroup.c:381]: Could this `for` be merged with the one directly above it? Would avoid iterating over the `pulseSpeed` array again.

File: motoman_driver/MotoPlus/Controller.c
Comment (gavanderhoorn) [inline motoman_driver/MotoPlus/Controller.c:665]: Just checking: if all `groupNo` are invalid, it immediately means that this function returns `FALSE`. I'm not sure a robot with all groups being invalid can be in motion, but this is on purpose, correct?

File: motoman_driver/MotoPlus/CtrlGroup.c
Comment (gavanderhoorn) [inline motoman_driver/MotoPlus/CtrlGroup.c:460]: style: the `memset(..)` above this one uses a plain, decimal, `0`.

File: motoman_driver/MotoPlus/IoServer.c
Comment (gavanderhoorn) [inline motoman_driver/MotoPlus/IoServer.c:100]: Is the `7` a nr with a special significance? Fi, `7` indicates the IoServer task could not be created?

If so: should we create constants for them, to avoid the magic nrs?

Comment (gavanderhoorn) [reply motoman_driver/MotoPlus/IoServer.c:100]: And the same comment for the other `mpSetAlarm(..)` invocations for task creation failure.

File: motoman_driver/MotoPlus/MotionServer.c
Comment (gavanderhoorn) [inline motoman_driver/MotoPlus/MotionServer.c:1559]: :+1:

File: motoman_driver/MotoPlus/MotionServer.c
Comment (gavanderhoorn) [inline motoman_driver/MotoPlus/MotionServer.c:1577]: This could be just an `if`, right (because of the `return`s)? Seeing an `else if` in isolation without an `else` always looks weird to me.

**Review by gavanderhoorn — COMMENTED** [review]
Haven't tested the new binaries yet, but here are some comments.

**Comment by gavanderhoorn** [comment]
Checked this on our SIA20F+linear rail.

This PR definitely improves the `in_motion` situation. `in_motion` stays `true` for the majority of the motion, and then goes to `false` three (3) `JointFeedback` messages before the robot has come to a complete stop. Joint velocity is also `> 0` for the three `JointFeedback` messages after `in_motion` has changed to `false`.

I haven't checked, but that could well be within the `START_MAX_PULSE_DEVIATION` that @EricMarcil mentions in https://github.com/ros-industrial/motoman/issues/219#issuecomment-400314030 (robot was moving rather slowly along the rail, and 3 msgs at 50 Hz is 75ms).

**Comment by gavanderhoorn** [comment]
@EricMarcil: I've added some comments.

File: motoman_driver/MotoPlus/CtrlGroup.c
Comment (EricMarcil) [reply motoman_driver/MotoPlus/CtrlGroup.c:381]: If you look closely, the axis correction makes reference to another source axis pulseSpeed.   Granted that the source axis is normally a previous axis but there is a possibility that it wouldn't.  So it is safer to calculate the pulseSpeed for all axes before doing the correction.

File: motoman_driver/MotoPlus/Controller.c
Comment (EricMarcil) [reply motoman_driver/MotoPlus/Controller.c:665]: That is correct and intentional.  If all groups are invalid, there really nothing to move.

File: motoman_driver/MotoPlus/IoServer.c
Comment (EricMarcil) [reply motoman_driver/MotoPlus/IoServer.c:100]: Yes, I've given each mpSetAlarm a different subcode (pretty much in the order they are created), so that if the alarm 8004 occurs, it could help identify which task wasn't started.  Generally speaking, I doubt that we would dig down to that level because whenever that 8004 occurs, it's probably because of parameters settings or you have multiple versions of the MotoRos running.  So I didn't go as far as creating constants for them.

File: motoman_driver/MotoPlus/MotionServer.c
Comment (EricMarcil) [reply motoman_driver/MotoPlus/MotionServer.c:1577]: Yes, it could.  That change was made by @ted-miller.  Personally, I tend to build the if, else if, else structure based on the conditions that is to be consider (in this case the qCnt value).  It's independently of what occurs after the condition is identified.  So I tend to agree with the way it is.

**Comment by gavanderhoorn** [comment]
@alemme / @andreaskoepf: do you see any chance of testing fe26038 perhaps on your SDA?

It would seem to address at least some of the comments by @andreaskoepf in #219.

**Comment by andreaskoepf** [comment]
@EricMarcil Great to see some improvements here!

@gavanderhoorn We will defiintely test this during the next two weeks. (Currently we are in final preparations for Rosvita v0 release (which hopefully will be the best or one of the best ROS GUIs for controlling the SDA). I plan to sync our Motoman driver with this repo in the July timeframe.)

**Comment by gavanderhoorn** [comment]
@andreaskoepf wrote:
> We will defiintely test this during the next two weeks. (Currently we are in final preparations for Rosvita v0 release (which hopefully will be the best or one of the best ROS GUIs for controlling the SDA). I plan to sync our Motoman driver with this repo in the July timeframe.)

Ok, good.

In the meantime I'll go ahead and merge this, as it solves some of the other issues we have with goals being flagged as completed too early.

If anything pops up during your testing please open a new issue.

**Comment by gavanderhoorn** [comment]
I'm not in the habit of doing this btw, but the test in this case would be in addition to the verification that has already been done by me and @ted-miller and @EricMarcil (at Motoman), and is also more a test to see whether the situation that you described in #219 has improved.

**Comment by gavanderhoorn** [comment]
@EricMarcil: as this PR was opened by me, I need one other maintainer to review this PR and approve it. Could you do that?

**Review by EricMarcil — APPROVED** [review]

**Comment by gavanderhoorn** [comment]
Thanks for the review @EricMarcil :+1:

**Pull request merged**
